### PR TITLE
[RFC] Refactor functions which find character in a string

### PR DESCRIPTION
**Pull request by ZyX-I**
Ref #1476

### CODE REVIEW

File: src/nvim/regexp.c [hunk 1]
```diff
@@ -6299,43 +6287,38 @@ static int cstrncmp(char_u *s1, char_u *s2, int *n)
 /*
  * cstrchr: This function is used a lot for simple searches, keep it fast!
  */
-static char_u *cstrchr(char_u *s, int c)
+static inline char_u *cstrchr(const char_u *const s, const int c)
+  FUNC_ATTR_PURE FUNC_ATTR_WARN_UNUSED_RESULT FUNC_ATTR_NONNULL_ALL
+  FUNC_ATTR_ALWAYS_INLINE
```
Comment (ZyX-I) [inline src/nvim/regexp.c]: This is “some work to avoid call overhead” removed from comments above.

**Comment by aktau** [comment]
> const are for telling that parameters will not be reassigned, not for forcing reassigning them. If needed, remove const. Though most of time const are fine because reassignings are either absent or have different semantics and are better be expressed via other variables.

Especially in the function parameter list, const is something only for the implementer of the function (if that). It basically muddles up the prototype of the function, making it less clear what's going on. For value types, this const adds nothing for the user, what do they care whether a parameter is `const int` or `int`? I'd much rather read the latter. Yet, in the same vein, I'd prefer reading `const char *str` over `char *str /* this function does not modify str */`. I also happen to think that in some cases, reassigning parameters is not a sin. Naming is hard, sometimes, you want to modify or replace a value (the parameter). It has a perfectly good name, you can either create `myname_fixed` or reassign `myname`. If you do the latter, you avoid two variables being live at the same time and the possible misuse of the old (unfixed) variant in later parts of the code. Relatedly, I'm not sure what things like `const int/size_t/...` are supposed to be solving.

> In any case, I like the idea of languages like Rust where const is default. In C const is not default, so I write const.

I like the idea of Rust as well, though I find it usually turns into letter soup for me (similar to what Mickens describes C++ to be like in [The Night Watch](http://scholar.harvard.edu/files/mickens/files/thenightwatch.pdf)). Perhaps I'm not clever enough for it.

> This function has alternatives, not early return. Also no other reasons for making if() if(). So no if() if(), your variant is less readable in my opinion and wastes more space for nothing useful.

I can see your point and won't belabour mine. This is all a bit subjective and perhaps I've been too exposed to Go (ref to advice: https://golang.org/doc/effective_go.html#control-structures).

What happens (in my head) is that when an `if` structure closes (and there's no else), I can throw the state away in my mind (or at least, it serializes better) (my mind goes: special case handled, let's continue). The following code is easier to read. In some cases, it's really more about alternative selection. I would say that in this case `c == 0` is more like a special/error case and `c < 0x80` is more like the alternative case.

File: src/nvim/strings.c [hunk 2]
```diff
@@ -401,54 +401,28 @@ int vim_strnicmp(const char *s1, const char *s2, size_t len)
 }
```
Comment (aktau) [inline src/nvim/strings.c]: The commit message should be: s/strstr/strchr/.

**Comment by ZyX-I** [comment]
Travis failure is test-single-includes hang up.

File: src/nvim/regexp.c [hunk 3]
```diff
@@ -6287,43 +6287,38 @@ static int cstrncmp(char_u *s1, char_u *s2, int *n)
...
+  // For UTF-8 need to use folded case.
+  if (c > 0x80) {
+    const int folded_c = utf_fold(c);
+    for (const char_u *p = s; *p != NUL; p += utfc_ptr2len(p)) {
+      if (utf_fold(utf_ptr2char(p)) == folded_c) {
+        return (char_u *)p;
+      }
+    }
+    return NULL;
+  }
```
Comment (aktau) [inline src/nvim/regexp.c]: This seems more like it's a bugfix instead of a refactor, no? If so, perhaps better to split the commits (for bisecting etc.)

Comment (ZyX-I) [inline src/nvim/regexp.c]: What a bugfix? There are no bug fixes here.

Comment (aktau) [inline src/nvim/regexp.c]: You're right, I got confused by the diff reordering. I thought this part was new.

**Comment by aktau** [comment]
LGTM

**Comment by justinmk** [comment]
Needs conflict resolution.

**Comment by ZyX-I** [comment]
Still using ascii functions after the merge: reasoning behind the change still applies as `mb_…` functions are as well still using libc locale-dependent functions.

**Comment by ZyX-I** [comment]
Looks like it could not find busted.

**Comment by ZyX-I** [comment]
Recreated merge commit using new master head.

**Comment by ZyX-I** [comment]
AppVeyor fail looks unrelated.

**Comment by ZyX-I** [comment]
QB failure is also unrelated: some of the tests merged to master uses execute() which is no longer there. Not sure why travis did not fail as well: most likely, locale is not supported there.

**Comment by ZyX-I** [comment]
Yes, travis reports unsupported locale.

**Comment by bfredl** [comment]
I also just noticed it. Should be no reason to not also run the Turkish tests on travis: #6495

**Comment by justinmk** [comment]
Is this ready?

**Comment by ZyX-I** [comment]
It is.

File: src/nvim/strings.c [hunk 4]
```diff
@@ -394,72 +394,28 @@ int vim_strnicmp(const char *s1, const char *s2, size_t len)
...
-    while (*p != NUL) {
-      int l = (*mb_ptr2len)(p);
-
-      // Avoid matching an illegal byte here.
-      if (l > 1 && utf_ptr2char(p) == c) {
-        return (char_u *) p;
-      }
-      p += l;
-    }
+  assert(c >= 0);
```
Comment (oni-link) [inline src/nvim/strings.c]: `:ls` and pressing `<F1>` while "Press ENTER ..." message is shown gives
```c
Program received signal SIGABRT, Aborted.
0x0000003102033315 in raise () from /lib64/libc.so.6
(gdb) bt
#0  0x0000003102033315 in raise () from /lib64/libc.so.6
#1  0x000000310203472a in abort () from /lib64/libc.so.6                                
#2  0x000000310202c337 in __assert_fail_base () from /lib64/libc.so.6          
#3  0x000000310202c3e2 in __assert_fail () from /lib64/libc.so.6                       
#4  0x0000000001701236 in vim_strchr (string=0x1a5e460 <.str.9> "\r\n ", c=-12651) at /home/oni-link/git/neovim/src/nvim/strings.c:439
#5  0x0000000000f88beb in wait_return (redraw=0) at /home/oni-link/git/neovim/src/nvim/message.c:935
#6  0x0000000000b2f266 in do_cmdline (cmdline=0x0, fgetline=0xc02db0 <getexline>, cookie=0x0, flags=0)
    at /home/oni-link/git/neovim/src/nvim/ex_docmd.c:955
#7  0x00000000010c7f83 in nv_colon (cap=0x7fffffffcfe0) at /home/oni-link/git/neovim/src/nvim/normal.c:4498
#8  0x00000000010afb00 in normal_execute (state=0x7fffffffcf40, key=58) at /home/oni-link/git/neovim/src/nvim/normal.c:1137
#9  0x00000000016fedbe in state_enter (s=0x7fffffffcf40) at /home/oni-link/git/neovim/src/nvim/state.c:61
#10 0x0000000001068140 in normal_enter (cmdwin=false, noexmode=false) at /home/oni-link/git/neovim/src/nvim/normal.c:467
#11 0x0000000000e43fc4 in main (argc=5, argv=0x7fffffffd828) at /home/oni-link/git/neovim/src/nvim/main.c:556
```

Comment (ZyX-I) [inline src/nvim/strings.c]: Given the next line it looks like this is a leftover of when I thought that c is never NUL.

Comment (ZyX-I) [inline src/nvim/strings.c]: Though no, it is >= 0. What -12651 is doing there?

Comment (teto) [inline src/nvim/strings.c]: on startup  a similar trace
```
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:58
#1  0x00007ffff6cfb3ea in __GI_abort () at abort.c:89
#2  0x00007ffff6cf1bb7 in __assert_fail_base (fmt=<optimized out>, 
    assertion=assertion@entry=0x5555558f1d06 "c >= 0", 
    file=file@entry=0x5555558f1cf0 "../src/nvim/strings.c", line=line@entry=439, 
    function=function@entry=0x5555558f2030 <__PRETTY_FUNCTION__.20488> "vim_strchr") at assert.c:92
#3  0x00007ffff6cf1c62 in __GI___assert_fail (assertion=0x5555558f1d06 "c >= 0", 
    file=0x5555558f1cf0 "../src/nvim/strings.c", line=439, 
    function=0x5555558f2030 <__PRETTY_FUNCTION__.20488> "vim_strchr") at assert.c:101
#4  0x00005555557d2ae8 in vim_strchr (string=0x5555558e4262 "\r\n ", c=-24829) at ../src/nvim/strings.c:439
#5  0x00005555556f60cf in wait_return (redraw=0) at ../src/nvim/message.c:935
#6  0x00005555556f8d45 in msg_end () at ../src/nvim/message.c:2441
#7  0x00005555556f48c0 in msg_attr_keep (s=0x555555b95660 <errbuf> "E121: Undefined variable: a:pat", 
    attr=3, keep=0) at ../src/nvim/message.c:194
```

Comment (ZyX-I) [inline src/nvim/strings.c]: @teto #6707

Comment (ZyX-I) [inline src/nvim/strings.c]: It appears that all special keys are intentionally encoded like this, see `#TERMCAP2KEY` macros.

**Pull request merged**